### PR TITLE
feat(dbt-ingestion): add documentation link from dbt source to institutionalMemory

### PR DESCRIPTION
**Pull request by ethan-cartwright**
### Description
We want the ability to add a link in the `+Add Link` section from the DBT source side. This PR adds dbt-side specified `documentation_link` as an institutionalMemory aspect on a dataset with the datahub-side `meta_mapping` specified `description` in the `config` (see the dbt.md file for details).

**NOTE: right now it replaces the current institution memory link with the new link. If we'd like the ability to add multiple links, or to check the dataset's current institutionalMemory and add our link to it, let me know.**

### Example
adds:
Screenshot 2023-08-22 at 4 52 19 PM

to:
Screenshot 2023-08-22 at 4 53 36 PM

### Testing
* unit tests
* local ingestion of dbt-cloud project [linked here](https://github.com/ethan-cartwright/sample-dbt/blob/main/models/base.yml#L16)

Results:
Screenshot 2023-08-22 at 4 54 26 PM

## Checklist

- [x] The PR conforms to DataHub's [Contributing Guideline](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md) (particularly [Commit Message Format](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md#commit-message-format))
- [x] Links to related issues (if applicable)
- [x] Tests for the changes have been added/updated (if applicable)
- [x] Docs related to the changes have been added/updated (if applicable). If a new feature has been added a Usage Guide has been added for the same.
- [x] For any breaking change/potential downtime/deprecation/big changes an entry has been made in [Updating DataHub](https://github.com/datahub-project/datahub/blob/master/docs/how/updating-datahub.md)

### CODE REVIEW

File: metadata-ingestion/examples/recipes/dbt_cloud_sample_to_datahub.dhub.yaml [hunk 1]
```diff
@@ -0,0 +1,16 @@
+source:
+    type: dbt-cloud
+    config:
+        project_id: '277462'
+        target_platform: postgres
+        stateful_ingestion:
+            enabled: false
+        account_id: '107298'
+        job_id: '399798'
```
Comment (hsheth2) [inline metadata-ingestion/examples/recipes/dbt_cloud_sample_to_datahub.dhub.yaml:9]: can you merge this stuff into the dbt sample recipe instead https://github.com/datahub-project/datahub/blob/05a0f3e2a67028249735e522c385df6306e01e7d/metadata-ingestion/docs/sources/dbt/dbt-cloud_recipe.yml#L12

File: metadata-ingestion/src/datahub/utilities/mapping.py [hunk 2]
```diff
@@ -12,6 +14,13 @@
     OwnershipTypeClass,
 )
 
+# Imports for metadata model classes
```
Comment (hsheth2) [inline metadata-ingestion/src/datahub/utilities/mapping.py:17]:
```suggestion
```

File: metadata-ingestion/src/datahub/utilities/mapping.py [hunk 3]
```diff
@@ -182,11 +194,41 @@ def convert_to_aspects(
...
+
         if Constants.ADD_TERM_OPERATION in operation_map:
             term_aspect = mce_builder.make_glossary_terms_aspect_from_urn_list(
                 sorted(operation_map[Constants.ADD_TERM_OPERATION])
             )
             aspect_map[Constants.ADD_TERM_OPERATION] = term_aspect
+
+        if Constants.ADD_DOC_LINK_OPERATION in operation_map:
+            try:
+                docs_dic = ast.literal_eval(
```
Comment (hsheth2) [inline metadata-ingestion/src/datahub/utilities/mapping.py:206]: ast eval feels strange here - could we just use `json.loads` or similar instead?

also, the ownership one supports ownership types without ever doing any parsing - can we use the same approach of a nested object here?

Comment (hsheth2) [inline metadata-ingestion/src/datahub/utilities/mapping.py:206]:
```suggestion
                docs_dict = ast.literal_eval(
```

File: metadata-ingestion/src/datahub/utilities/mapping.py [hunk 4]
```diff
@@ -182,11 +194,41 @@ def convert_to_aspects(
...
+                    operation_map[Constants.ADD_DOC_LINK_OPERATION].pop()
+                )
+
+                now = int(time.time() * 1000)  # milliseconds since epoch
+                institutional_memory_element = InstitutionalMemoryMetadataClass(
+                    url=docs_dic["link"],
+                    description=docs_dic["descr"],
```
Comment (hsheth2) [inline metadata-ingestion/src/datahub/utilities/mapping.py:213]: descr -> description

**Review by hsheth2 — COMMENTED** [review]
some initial comments from a first pass

we'll need docs for this in the dbt.md file too

File: metadata-ingestion/src/datahub/utilities/mapping.py (hunk 3 above)
Comment (ethan-cartwright) [reply metadata-ingestion/src/datahub/utilities/mapping.py]: good call, went with the second option. Parsing was indeed unnecessary

File: metadata-ingestion/docs/sources/dbt/dbt-cloud_recipe.yml [hunk 5]
```diff
@@ -6,9 +6,9 @@ source:
     # In the URL https://cloud.getdbt.com/next/deploy/107298/projects/175705/jobs/148094,
     # 107298 is the account_id, 175705 is the project_id, and 148094 is the job_id
 
-    account_id: # set to your dbt cloud account id
-    project_id: # set to your dbt cloud project id
-    job_id: # set to your dbt cloud job id
+    account_id: '107298' # set to your dbt cloud account id
```
Comment (hsheth2) [inline metadata-ingestion/docs/sources/dbt/dbt-cloud_recipe.yml:9]: make it clear that this is not something they can just copy
```suggestion
    account_id: "${DBT_ACCOUNT_ID}" # set to your dbt cloud account id
```

File: metadata-ingestion/docs/sources/dbt/dbt.md [hunk 6]
```diff
@@ -38,6 +38,12 @@ meta_mapping:
     operation: "add_terms"
     config:
       separator: ","
+  documentation_link:
+    match: ".*"
+    operation: "add_link"
+    config:
+      link: "https://en.wikipedia.org/wiki/Anchor_text"
```
Comment (hsheth2) [inline metadata-ingestion/docs/sources/dbt/dbt.md:45]: let's change this example to use a pattern

in reality, we don't expect people to just add the same link to every asset. instead, the link is going to be in their meta section and we have to extract it out

File: metadata-ingestion/src/datahub/utilities/mapping.py [hunk 7]
```diff
@@ -182,11 +192,39 @@ def convert_to_aspects(
...
+
         if Constants.ADD_TERM_OPERATION in operation_map:
             term_aspect = mce_builder.make_glossary_terms_aspect_from_urn_list(
                 sorted(operation_map[Constants.ADD_TERM_OPERATION])
             )
             aspect_map[Constants.ADD_TERM_OPERATION] = term_aspect
+
+        if Constants.ADD_DOC_LINK_OPERATION in operation_map:
+            try:
+                docs_dict = operation_map[Constants.ADD_DOC_LINK_OPERATION].pop()
```
Comment (hsheth2) [inline metadata-ingestion/src/datahub/utilities/mapping.py:204]: using pop() here doesn't seem right

File: metadata-ingestion/tests/unit/test_mapping.py [hunk 8]
```diff
@@ -231,3 +232,43 @@ def test_operation_processor_advanced_matching_tags():
...
+    processor = OperationProcessor(
+        operation_defs={
+            "documentation_link": {
+                "match": ".*",
+                "operation": "add_doc_link",
+                "config": {"link": "{{ $match }}"},
+            },
+        },
+    )
+    # we require a description, so this should stay empty
```
Comment (hsheth2) [inline metadata-ingestion/tests/unit/test_mapping.py:272]: if description is missing, do we issue a generic error message or a detailed one?

File: metadata-ingestion/tests/unit/test_mapping.py [hunk 9]
```diff
@@ -231,3 +232,43 @@ def test_operation_processor_advanced_matching_tags():
...
+
+
+def test_operation_processor_institutional_memory_no_description():
+    raw_props = {
+        "documentation_link": "test.com",
+    }
+    processor = OperationProcessor(
+        operation_defs={
+            "documentation_link": {
+                "match": ".*",
```
Comment (hsheth2) [inline metadata-ingestion/tests/unit/test_mapping.py:266]: can we also test that match param works correctly?

Comment (hsheth2) [reply metadata-ingestion/tests/unit/test_mapping.py:266]: something similar to the test case here https://github.com/datahub-project/datahub/blob/5b9fd977eb60d4a1a41d6aa7cadfb7bce01a93ab/metadata-ingestion/tests/integration/dbt/test_dbt.py#L62

e.g. match=`(.+)(?:#.*)?` and raw props =`http://example.com/my-page#stuff-to-ignore` should produce an aspect with link = `http://example.com/my-page`

**Review by hsheth2 — APPROVED** [review]

**Pull request merged**